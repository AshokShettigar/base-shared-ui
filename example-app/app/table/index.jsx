import { CheckBox, Input, Table } from 'platform-common-ui';
import BaseComponentView from 'app/components/base-component-view';

const Cell = Table.Cell;
const CheckBoxColumn = Table.CheckBoxColumn;
const Column = Table.Column;
const Row = Table.Row;
const SelectRowCell = Table.SelectRowCell;

// eslint-disable-next-line no-alert
const onColumnClickHandler = id => alert(`Column: ${id} clicked!`);

export default class TableExample extends BaseComponentView {
    constructor(props) {
        super(props);
        this.state = {
            height: 200,
            isAllSelected: false,
            isShowFooter: true,
            isShowSelectColumn: true,
            numCols: 3,
            numRows: 3,
            selectedRows: {}
        };

        this.onFixHeightCheckBoxChange =
            () => this.setState({ height: this.state.height ? undefined : 200 });
        this.onHeightChange = value => {
            const height = parseInt(value, 10);
            this.setState({ height });
        };
        this.onNumColsChange = numCols => this.setState({ numCols });
        this.onNumRowsChange = numRows => this.setState({
            numRows,
            isAllSelected: false,
            selectedRows: {}
        });
        this.onSelectRow = index => () => {
            const selectedRows = Object.assign({}, this.state.selectedRows);
            selectedRows[index] = !this.state.selectedRows[index];
            this.setState({ selectedRows });
        };
        this.onShowFooterCheckBoxChange = () => {
            this.setState({ isShowFooter: !this.state.isShowFooter });
            setTimeout(() => this.table.adjustHeight(), 0);
        };
        this.onShowSelectColumnChange =
            () => this.setState({ isShowSelectColumn: !this.state.isShowSelectColumn });
        this.toggleSelectAll = () => {
            if (this.state.isAllSelected) {
                this.setState({
                    isAllSelected: false,
                    selectedRows: {}
                });
                return;
            }

            const selectedRows = {};
            for (let i = 0; i < this.state.numRows; i++) {
                selectedRows[i] = true;
            }
            this.setState({
                selectedRows,
                isAllSelected: true
            });
        };
    }

    getOptions() {
        return (
            <div>
                <Input
                    id="heightInput"
                    label="Table Height"
                    onChange={this.onHeightChange}
                    value={this.state.height || ''}
                >
                    <CheckBox
                        id="ckFixHeight"
                        isChecked={!!this.state.height}
                        onChange={this.onFixHeightCheckBoxChange}
                        value="fixHeight"
                    >
                        Fixed height
                    </CheckBox>
                </Input>
                <Input
                    id="numColInput"
                    label="Number of Columns"
                    onChange={this.onNumColsChange}
                    value={this.state.numCols}
                />
                <Input
                    id="numRowInput"
                    label="Number of Rows"
                    onChange={this.onNumRowsChange}
                    value={this.state.numRows}
                />
                <CheckBox
                    id="ckShowFooter"
                    isChecked={this.state.isShowFooter}
                    onChange={this.onShowFooterCheckBoxChange}
                    value="footer"
                >
                    Show Footer
                </CheckBox>
                <CheckBox
                    id="ckShowSelect"
                    isChecked={this.state.isShowSelectColumn}
                    onChange={this.onShowSelectColumnChange}
                    value="footer"
                >
                    Show Select Check Box Column
                </CheckBox>
            </div>
        );
    }

    getColumns() {
        const cols = [];
        if (this.state.isShowSelectColumn) {
            cols.push(
                <CheckBoxColumn
                    isAllSelected={this.state.isAllSelected}
                    key="col_ck"
                    onClick={this.toggleSelectAll}
                />
            );
        }

        for (let i = 0; i < this.state.numCols; i++) {
            cols.push(
                <Column
                    id={`col_${i}`}
                    isSortable
                    key={`col_${i}`}
                    onClick={onColumnClickHandler}
                    sortDirection={Table.SORT_DIRECTION.ASC}
                >
                    {`Header ${i + 1}`}
                </Column>
            );
        }
        return cols;
    }

    getRows() {
        const rows = [];
        for (let i = 0; i < this.state.numRows; i++) {
            const cells = [];
            if (this.state.isShowSelectColumn) {
                cells.push(
                    <SelectRowCell
                        id={`cb_${i}`}
                        isSelected={this.state.selectedRows[i]}
                        key="c_select"
                        onChange={this.onSelectRow(i)}
                    />
                );
            }
            for (let j = 0; j < this.state.numCols; j++) {
                cells.push(
                    <Cell key={`c_${j}`}>
                        {`Data ${j + 1}`}
                    </Cell>
                );
            }
            rows.push(
                <Row
                    isSelected={this.state.selectedRows[i]}
                    key={`r_${i}`}
                >
                    {cells}
                </Row>
            );
        }
        return rows;
    }

    getFooter() {
        if (!this.state.isShowFooter) return false;
        return (<div>Footer</div>);
    }

    renderComponent() {
        return (
            <Table
                columns={this.getColumns()}
                footer={this.getFooter()}
                height={this.state.height}
                ref={c => { this.table = c; }}
            >
                {this.getRows()}
            </Table>
        );
    }
}
