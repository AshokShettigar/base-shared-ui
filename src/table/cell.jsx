import PropTypes from 'prop-types';
import classNames from 'classnames';

const Cell = props => {
    const className = classNames('cn-table__cell', props.className);
    return (
        <td className={className}>
            {props.children}
        </td>
    );
};

Cell.displayName = 'TableCell';

Cell.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default Cell;
