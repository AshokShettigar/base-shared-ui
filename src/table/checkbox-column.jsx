import PropTypes from 'prop-types';
import Column from 'platform-common-ui/table/column';
import { Icon } from 'platform-common-ui';

const CheckBoxColumn = props => {
    const icon = props.isAllSelected ? 'check-all' : 'check-none';
    return (
        <Column id={props.id} onClick={props.onClick} width="38px">
            <Icon className="cn-table__header-icon" icon={icon} />
        </Column>
    );
};

CheckBoxColumn.displayName = 'CheckBoxColumn';

CheckBoxColumn.propTypes = {
    id: PropTypes.string.isRequired,
    isAllSelected: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired
};

CheckBoxColumn.defaultProps = {
    id: 'checkbox_col',
    isAllSelected: false
};

export default CheckBoxColumn;
