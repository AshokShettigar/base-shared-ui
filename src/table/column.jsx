import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Icon } from 'platform-common-ui';

export const SORT_DIRECTION = {
    ASC: 'asc',
    DESC: 'desc'
};

const SORT_ICON = {
    [SORT_DIRECTION.ASC]: 'arrow-up',
    [SORT_DIRECTION.DESC]: 'arrow-down'
};

const onClickHandler = (callback, id) => () => callback(id);

const Column = props => {
    const className = classNames({
        'cn-table__column': true,
        'cn-table__column--is-sortable': props.isSortable
    }, props.className);
    const style = {
        maxWidth: props.width || props.minWidth || 100,
        minWidth: props.width || props.minWidth || 100
    };
    if (props.width) {
        style.width = props.width;
    }

    /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
    /* eslint-disable jsx-ally/no-noninteractive-element-to-interactive-role */
    return (
        <th
            className={className}
            onClick={onClickHandler(props.onClick, props.id)}
            style={style}
        >
            {props.children}
            {props.sortDirection &&
                <Icon
                    className="cn-table__column__sort-icon"
                    icon={SORT_ICON[props.sortDirection]}
                />
            }
        </th>
    );
    /* eslint-disable jsx-ally/no-noninteractive-element-to-interactive-role */
    /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
};

Column.displayName = 'TableColumn';

Column.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    id: PropTypes.string.isRequired,
    isSortable: PropTypes.bool.isRequired,
    minWidth: PropTypes.string,
    onClick: PropTypes.func,
    sortDirection: PropTypes.oneOf([SORT_DIRECTION.ASC, SORT_DIRECTION.DESC]),
    width: PropTypes.string
};

Column.defaultProps = {
    isSortable: false
};

export default Column;
