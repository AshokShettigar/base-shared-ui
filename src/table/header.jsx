import { cloneElement } from 'react';
import PropTypes from 'prop-types';

function getWidth(widths, position) {
    if (widths.length) {
        return `${widths[position]}px`;
    }

    return '100px';
}

const Header = props => (
    <table className="cn-table__header" style={{ left: props.scrollLeft }}>
        <thead>
            <tr>
                {props.columns.map((column, i) => cloneElement(
                    column,
                    { width: getWidth(props.widths, i) }
                ))}
            </tr>
        </thead>
    </table>
);

Header.displayName = 'TableHeader';

Header.propTypes = {
    columns: PropTypes.node.isRequired,
    scrollLeft: PropTypes.number.isRequired,
    widths: PropTypes.arrayOf(PropTypes.number).isRequired
};

Header.defaultProps = {
    scrollLeft: 0,
    widths: []
};

export default Header;
