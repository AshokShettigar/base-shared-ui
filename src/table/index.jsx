import { Component } from 'react';
import PropTypes from 'prop-types';
import Cell from 'platform-common-ui/table/cell';
import CheckBoxColumn from 'platform-common-ui/table/checkbox-column';
import Column, { SORT_DIRECTION } from 'platform-common-ui/table/column';
import Header from 'platform-common-ui/table/header';
import Row from 'platform-common-ui/table/row';
import SelectRowCell from 'platform-common-ui/table/select-row-cell';

const BOTTOM_MARGIN = 20;

// eslint-disable-next-line import/prefer-default-export
export class Table extends Component {
    constructor(props) {
        super(props);
        this.state = {
            height: 300,
            left: 0,
            top: -36,
            columnWidths: []
        };

        this.resizeHandler = () =>
            window.requestAnimationFrame(() => {
                this.adjustHeight(this.props.height);
                this.adjustColumnWidths();
            });
        this.syncScrollHandler = () => window.requestAnimationFrame(() => this.syncScroll());
    }

    componentDidMount() {
        // make sure that the immediate instructions finish first to draw the component
        // on screen before calling to initailize
        window.setTimeout(() => this.initialize(), 0);
        window.addEventListener('resize', this.resizeHandler);
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.height !== nextProps.height) {
            this.adjustHeight(nextProps.height);
        }
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevState === this.state
            && (this.props.columns !== prevProps.columns
            || this.props.children !== prevProps.children)) {
            // if the component was updated via properties and columns are different,
            // then adjust the column widths
            this.adjustColumnWidths();
        }
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.resizeHandler);
    }

    adjustHeight(height) {
        if (height) {
            this.setState({ height: this.props.height });
        }

        const footerHeight = this.footer ? this.footer.getBoundingClientRect().height : 0;
        const newHeight = window.document.body.clientHeight
            - this.tableBody.getBoundingClientRect().top
            - footerHeight
            - BOTTOM_MARGIN;
        this.setState({ height: newHeight });
    }

    adjustColumnWidths() {
        const columnWidths = [];
        // can't use map function on children array
        for (let i = 0; i < this.hiddenHeaders.children.length; i++) {
            columnWidths.push(this.hiddenHeaders.children[i].clientWidth);
        }
        this.setState({ columnWidths });
    }

    initialize() {
        this.initializeTableBodyTop();
        this.adjustHeight(this.props.height);
        this.adjustColumnWidths();
    }

    initializeTableBodyTop() {
        const top = 1 - this.hiddenHeaders.getBoundingClientRect().height;
        this.setState({ top });
    }

    syncScroll() {
        if (this.props.onScroll) {
            this.props.onScroll();
        }
        this.setState({ left: -this.tableBody.scrollLeft });
    }

    renderFooter() {
        return (
            <div className="cn-table__footer" ref={footer => { this.footer = footer; }}>
                {this.props.footer}
            </div>
        );
    }

    render() {
        const tableBodyStyle = {
            height: this.state.height
        };

        return (
            <div className="cn-table">
                <Header
                    columns={this.props.columns}
                    scrollLeft={this.state.left}
                    widths={this.state.columnWidths}
                />
                <div
                    className="cn-table__body"
                    onScroll={this.syncScrollHandler}
                    ref={tableBody => { this.tableBody = tableBody; }}
                    style={tableBodyStyle}
                >
                    <div
                        className="cn-table__body__container"
                    >
                        <table
                            className="cn-table__rows"
                            style={{ top: this.state.top }}
                        >
                            <thead className="cn-table__header">
                                <tr ref={headers => { this.hiddenHeaders = headers; }}>
                                    {this.props.columns}
                                </tr>
                            </thead>
                            <tbody>
                                {this.props.children}
                            </tbody>
                        </table>
                    </div>
                </div>
                {this.props.footer && this.renderFooter()}
            </div>
        );
    }
}

Table.propTypes = {
    children: PropTypes.node.isRequired,
    footer: PropTypes.node,
    columns: PropTypes.node.isRequired,
    height: PropTypes.number,
    onScroll: PropTypes.func
};

Table.SORT_DIRECTION = SORT_DIRECTION;
Table.Cell = Cell;
Table.CheckBoxColumn = CheckBoxColumn;
Table.Column = Column;
Table.Row = Row;
Table.SelectRowCell = SelectRowCell;
