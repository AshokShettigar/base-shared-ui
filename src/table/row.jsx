import classNames from 'classnames';
import PropTypes from 'prop-types';

const Row = props => {
    const className = classNames({
        'cn-table__row': true,
        'cn-table__row--selected': props.isSelected
    }, props.className);
    return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <tr className={className} onClick={props.onClick}>
            {props.children}
        </tr>
    );
};

Row.displayName = 'TableRow';

Row.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    isSelected: PropTypes.bool.isRequired,
    onClick: PropTypes.func
};

Row.defaultProps = {
    isSelected: false
};

export default Row;
