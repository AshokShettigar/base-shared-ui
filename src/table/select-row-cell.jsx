import PropTypes from 'prop-types';
import Cell from 'platform-common-ui/table/cell';
import { CheckBox } from 'platform-common-ui/checkbox';

const SelectRowCell = props => (
    <Cell width={36}>
        <CheckBox
            hideLabel
            id={props.id}
            isChecked={props.isSelected}
            onChange={props.onChange}
            value="row"
        >
            Select row
        </CheckBox>
    </Cell>
);

SelectRowCell.displayName = 'SelectRowCell';

SelectRowCell.propTypes = {
    id: PropTypes.string.isRequired,
    isSelected: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired
};

SelectRowCell.defaultProps = {
    isSelected: false
};

export default SelectRowCell;
