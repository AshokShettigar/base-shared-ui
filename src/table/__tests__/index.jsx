import { shallow, mount } from 'enzyme';
import { Table } from 'platform-common-ui/table';

describe('table/index', function() {
    const origAddEventListener = window.addEventListener;
    const origRemoveEventListener = window.removeEventListener;
    const origRequestAnimationFrame = window.requestAnimationFrame;
    const origSetTimeout = window.setTimeout;
    const origAdjustColumnWidths = Table.prototype.adjustColumnWidths;
    const origAdjustHeight = Table.prototype.adjustHeight;
    const origInitializeTableBodyTop = Table.prototype.initializeTableBodyTop;
    const origInitialize = Table.prototype.initialize;
    const origSyncScroll = Table.prototype.syncScroll;
    const origSetState = Table.prototype.setState;

    const columns = [
        <Table.Column id="1" key="1">Head 1</Table.Column>,
        <Table.Column id="2" key="2">Head 2</Table.Column>
    ];

    const data = [(
        <Table.Row key="1">
            <Table.Cell>Data 1-1</Table.Cell>
            <Table.Cell>Data 1-2</Table.Cell>
        </Table.Row>
    ), (
        <Table.Row key="2">
            <Table.Cell>Data 2-1</Table.Cell>
            <Table.Cell>Data 2-2</Table.Cell>
        </Table.Row>
    )];

    afterEach(function() {
        window.addEventListener = origAddEventListener;
        window.removeEventListener = origRemoveEventListener;
        window.requestAnimationFrame = origRequestAnimationFrame;
        window.setTimeout = origSetTimeout;
        Table.prototype.adjustColumnWidths = origAdjustColumnWidths;
        Table.prototype.adjustHeight = origAdjustHeight;
        Table.prototype.initialize = origInitialize;
        Table.prototype.initializeTableBodyTop = origInitializeTableBodyTop;
        Table.prototype.syncScroll = origSyncScroll;
        Table.prototype.setState = origSetState;
    });

    it('should render the component', function() {
        const wrapper = shallow(
            <Table columns={columns}>{data}</Table>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the component with height', function() {
        const wrapper = shallow(
            <Table columns={columns} height={500}>{data}</Table>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the component with footer', function() {
        const footer = <div>Footer</div>;
        const wrapper = shallow(
            <Table columns={columns} footer={footer}>{data}</Table>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should call did mount', function() {
        window.addEventListener = jest.fn();
        window.setTimeout = jest.fn(init => init());
        Table.prototype.initialize = jest.fn();
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        const resizeHandler = wrapper.instance().resizeHandler;
        expect(window.addEventListener).toHaveBeenCalledWith('resize', resizeHandler);
        expect(window.setTimeout).toHaveBeenCalledWith(expect.any(Function), 0);
        expect(Table.prototype.initialize).toHaveBeenCalledTimes(1);
    });

    it('should call did mount with footer', function() {
        window.addEventListener = jest.fn();
        window.setTimeout = jest.fn();
        const wrapper = mount(
            <Table columns={columns} footer="Footer">{data}</Table>
        );
        const resizeHandler = wrapper.instance().resizeHandler;
        expect(window.addEventListener).toHaveBeenCalledWith('resize', resizeHandler);
        expect(window.setTimeout).toHaveBeenCalledWith(expect.any(Function), 0);
    });

    it('should call initialize', function() {
        Table.prototype.adjustColumnWidths = jest.fn();
        Table.prototype.adjustHeight = jest.fn();
        Table.prototype.initializeTableBodyTop = jest.fn();
        const table = new Table({ height: 10 });
        table.initialize();
        expect(Table.prototype.adjustHeight).toHaveBeenCalledWith(10);
        expect(Table.prototype.adjustColumnWidths).toHaveBeenCalledTimes(1);
        expect(Table.prototype.initializeTableBodyTop).toHaveBeenCalledTimes(1);
    });

    it('should remove event listener on unmount', function() {
        window.removeEventListener = jest.fn();
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        const resizeHandler = wrapper.instance().resizeHandler;
        wrapper.unmount();
        expect(window.removeEventListener).toHaveBeenCalledWith('resize', resizeHandler);
    });

    it('should adjust the height after the height change', function() {
        Table.prototype.adjustHeight = jest.fn();
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        wrapper.setProps({ height: 500 });
        expect(Table.prototype.adjustHeight).toHaveBeenCalledWith(500);
    });

    it('should adjust the height with the footer', function() {
        Table.prototype.setState = jest.fn();
        const table = new Table();
        table.footer = {
            getBoundingClientRect: () => ({ height: 5 })
        };
        table.tableBody = {
            getBoundingClientRect: () => ({ top: 10 })
        };
        table.adjustHeight();
        expect(Table.prototype.setState).toHaveBeenCalledWith({
            height: window.document.body.clientHeight - 35
        });
    });

    it('should adjust the height without the footer', function() {
        Table.prototype.setState = jest.fn();
        const table = new Table();
        table.tableBody = {
            getBoundingClientRect: () => ({ top: 10 })
        };
        table.adjustHeight();
        expect(Table.prototype.setState).toHaveBeenCalledWith({
            height: window.document.body.clientHeight - 30
        });
    });

    it('should not call adjust the height if the height is same', function() {
        const wrapper = mount(
            <Table columns={columns} height={500} >{data}</Table>
        );
        Table.prototype.adjustHeight = jest.fn();
        wrapper.setProps({ height: 500 });
        expect(Table.prototype.adjustHeight).toHaveBeenCalledTimes(0);
    });

    it('should call resize handler on window resize', function() {
        window.requestAnimationFrame = function(callback) {
            callback();
        };
        Table.prototype.adjustHeight = jest.fn();
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        window.dispatchEvent(new Event('resize'));
        expect(wrapper.instance().adjustHeight).toHaveBeenCalled();
    });

    it('should call sync scroll handler on scroll', function() {
        window.requestAnimationFrame = function(callback) {
            callback();
        };
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        const instance = wrapper.instance();
        Table.prototype.setState = jest.fn();
        instance.tableBody = { scrollLeft: 20 };
        wrapper.find('.cn-table__body').simulate('scroll');
        expect(instance.setState).toHaveBeenCalledWith({ left: -20 });
    });

    it('should call custom onScroll function when it scrolls', function() {
        window.requestAnimationFrame = function(callback) {
            callback();
        };
        const onScrollHandler = jest.fn();
        const wrapper = mount(
            <Table columns={columns} onScroll={onScrollHandler}>{data}</Table>
        );
        const instance = wrapper.instance();
        Table.prototype.setState = jest.fn();
        instance.tableBody = { scrollLeft: 20 };
        wrapper.find('.cn-table__body').simulate('scroll');
        expect(onScrollHandler).toHaveBeenCalledTimes(1);
    });

    it('should initialize table body top', function() {
        Table.prototype.setState = jest.fn();
        const table = new Table();
        table.hiddenHeaders = {
            getBoundingClientRect: () => ({ height: 20 })
        };
        table.initializeTableBodyTop();
        expect(Table.prototype.setState).toHaveBeenCalledWith({ top: -19 });
    });

    it('should call adjustColumnWidths on columns property update', function() {
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        Table.prototype.adjustColumnWidths = jest.fn();
        wrapper.setProps({ columns: [] });
        expect(Table.prototype.adjustColumnWidths).toHaveBeenCalledTimes(1);
    });

    it('should call adjustColumnWidths on columns property update', function() {
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        Table.prototype.adjustColumnWidths = jest.fn();
        wrapper.setProps({ columns: [] });
        expect(Table.prototype.adjustColumnWidths).toHaveBeenCalledTimes(1);
    });

    it('should not call adjustColumnWidths on other property or state changes', function() {
        const wrapper = mount(
            <Table columns={columns}>{data}</Table>
        );
        Table.prototype.adjustColumnWidths = jest.fn();
        wrapper.setProps({ height: 100 });
        expect(Table.prototype.adjustColumnWidths).toHaveBeenCalledTimes(0);
        wrapper.setState({});
        expect(Table.prototype.adjustColumnWidths).toHaveBeenCalledTimes(0);
    });

    it('should test adjustColumnWidths', function() {
        Table.prototype.setState = jest.fn();
        const table = new Table();
        table.hiddenHeaders = {
            children: [
                { clientWidth: 5 },
                { clientWidth: 6 },
                { clientWidth: 7 }
            ]
        };
        table.adjustColumnWidths();
        expect(Table.prototype.setState).toHaveBeenCalledWith({ columnWidths: [5, 6, 7] });
    });
});
