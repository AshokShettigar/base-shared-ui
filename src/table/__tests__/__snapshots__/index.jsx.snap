// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table/index should render the component 1`] = `
<div
  className="cn-table"
  data-component="Table"
>
  <TableHeader
    columns={
      Array [
        <TableColumn
          id="1"
          isSortable={false}
      >
          Head 1
      </TableColumn>,
        <TableColumn
          id="2"
          isSortable={false}
      >
          Head 2
      </TableColumn>,
      ]
    }
    scrollLeft={0}
    widths={Array []}
  />
  <div
    className="cn-table__body"
    onScroll={[Function]}
    style={
      Object {
        "height": 300,
      }
    }
  >
    <div
      className="cn-table__body__container"
    >
      <table
        className="cn-table__rows"
        style={
          Object {
            "top": -36,
          }
        }
      >
        <thead
          className="cn-table__header"
        >
          <tr>
            <TableColumn
              id="1"
              isSortable={false}
            >
              Head 1
            </TableColumn>
            <TableColumn
              id="2"
              isSortable={false}
            >
              Head 2
            </TableColumn>
          </tr>
        </thead>
        <tbody>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 1-1
            </TableCell>
            <TableCell>
              Data 1-2
            </TableCell>
          </TableRow>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 2-1
            </TableCell>
            <TableCell>
              Data 2-2
            </TableCell>
          </TableRow>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`table/index should render the component with footer 1`] = `
<div
  className="cn-table"
  data-component="Table"
>
  <TableHeader
    columns={
      Array [
        <TableColumn
          id="1"
          isSortable={false}
      >
          Head 1
      </TableColumn>,
        <TableColumn
          id="2"
          isSortable={false}
      >
          Head 2
      </TableColumn>,
      ]
    }
    scrollLeft={0}
    widths={Array []}
  />
  <div
    className="cn-table__body"
    onScroll={[Function]}
    style={
      Object {
        "height": 300,
      }
    }
  >
    <div
      className="cn-table__body__container"
    >
      <table
        className="cn-table__rows"
        style={
          Object {
            "top": -36,
          }
        }
      >
        <thead
          className="cn-table__header"
        >
          <tr>
            <TableColumn
              id="1"
              isSortable={false}
            >
              Head 1
            </TableColumn>
            <TableColumn
              id="2"
              isSortable={false}
            >
              Head 2
            </TableColumn>
          </tr>
        </thead>
        <tbody>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 1-1
            </TableCell>
            <TableCell>
              Data 1-2
            </TableCell>
          </TableRow>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 2-1
            </TableCell>
            <TableCell>
              Data 2-2
            </TableCell>
          </TableRow>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="cn-table__footer"
  >
    <div>
      Footer
    </div>
  </div>
</div>
`;

exports[`table/index should render the component with height 1`] = `
<div
  className="cn-table"
  data-component="Table"
>
  <TableHeader
    columns={
      Array [
        <TableColumn
          id="1"
          isSortable={false}
      >
          Head 1
      </TableColumn>,
        <TableColumn
          id="2"
          isSortable={false}
      >
          Head 2
      </TableColumn>,
      ]
    }
    scrollLeft={0}
    widths={Array []}
  />
  <div
    className="cn-table__body"
    onScroll={[Function]}
    style={
      Object {
        "height": 300,
      }
    }
  >
    <div
      className="cn-table__body__container"
    >
      <table
        className="cn-table__rows"
        style={
          Object {
            "top": -36,
          }
        }
      >
        <thead
          className="cn-table__header"
        >
          <tr>
            <TableColumn
              id="1"
              isSortable={false}
            >
              Head 1
            </TableColumn>
            <TableColumn
              id="2"
              isSortable={false}
            >
              Head 2
            </TableColumn>
          </tr>
        </thead>
        <tbody>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 1-1
            </TableCell>
            <TableCell>
              Data 1-2
            </TableCell>
          </TableRow>
          <TableRow
            isSelected={false}
          >
            <TableCell>
              Data 2-1
            </TableCell>
            <TableCell>
              Data 2-2
            </TableCell>
          </TableRow>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
